### DISCUSSION TITLE
Common code between plugins

### DISCUSSION THREAD

**Discussion by andrewgoz** [TiddlyWiki/TiddlyWiki5#7682] [Q&A]
All the plugins I've seen are structured so their tiddlers are named like `$:/plugins/pluginname/tiddlernames`.

What if I want to develop multiple plugins that have some common code? I'd like to maintain that common code in one spot, but still have each plugin be independently installable.

What options do I have?

**Comment by pmario** [comment]
You can have a closer look at the codemirror plugin(s). 

- The main library is defined at: [codemirror plugin.info](https://github.com/Jermolene/TiddlyWiki5/blob/6c67dc82356355234292da4b0034c4019c0e0fb8/plugins/tiddlywiki/codemirror/plugin.info)
- The [autocomplete addon plugin.info](https://github.com/Jermolene/TiddlyWiki5/blob/6c67dc82356355234292da4b0034c4019c0e0fb8/plugins/tiddlywiki/codemirror-autocomplete/plugin.info) has more meta data

There is the `"parent-plugin": "$:/plugins/tiddlywiki/codemirror",` which "links" to the main library. 

- This parameter used to import the parent if you only would import the "sub-plugin"
- This parameter is also used in the core plugin import UI. 

![image](https://github.com/Jermolene/TiddlyWiki5/assets/374655/0642acbc-4ec5-4d3d-b295-df77cd694d39)

There is the `"dependents": ["$:/plugins/tiddlywiki/codemirror-mode-css"],` parameter, which is an array. Eg: The codemirror-autocomplete plugin relies on the codemirror-mode-css plugin

- It allows do define other plugins, that this plugin relies on. 

That's basically it. It should give you everything you need to have very basic dependency resolution.

**Reply by andrewgoz** [reply]
Thank you. That is certainly something I should study to learn more.

However, I've reconsidered my approach. I think trying to link a common library to all the plugins is a bad idea, especially if I make a change and it propagates to all the plugins and breaks one.

I think I'll put all my common code into one tiddler, then copy-and-paste between them as required to keep the plugins up-to-date. I'm not planning to do very many related plugins, so it should be managable.

**Reply by linonetwo** [reply]
then you can use soft link to auto copy the file
